doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1446 @@
+    SNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+
+COUNTERS_NAPT_TWICE:ip_protocol:src_ip:src_l4_port:dst_ip:dst_l4_port
+    DNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    DNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+    SNAT_TRANSLATIONS_PKTS   : <packets_counter_value>
+    SNAT_TRANSLATIONS_BYTES  : <bytes_counter_value>
+```
+
+## 3.3 Switch State Service Design
+Following changes are done in the orchagent. 
+
+### 3.3.1 NatMgr daemon
+ NatMgrd gets the STATIC_NAPT, STATIC_NAT, NAT_POOL, NAT_GLOBAL, NAT_BINDINGS config changes from CONFIG_DB.
+ NatMgr is responsible for pushing the Static NAT/NAPT entries and the NAT_GLOBAL configuration into the APP_DB. It also programs the Static NAT/NAPT entries and the NAT_POOL to ACL binding configuration as iptable rules in the kernel.
+
+Before acting upon the Static NAPT configuration, NatMgrd checks with the STATE_DB that the matching global IP interface is configured in the system (state == ok). 
+
+For a STATIC_NAPT entry and the interface configuration as below: 
+```
+STATIC_NAPT|65.55.42.1|TCP|1024 
+    "local_ip": 20.0.0.1
+    "local_port" :6000 
+    "nat_type": "dnat"
+
+INTERFACE|Ethernet15|65.55.42.1/24
+...
+
+PORT|Ethernet15
+    "nat_zone": 1
+
+```
+the following iptable rules are added for inbound and outbound directions in the nat table as below:
+```
+iptables -t nat -A PREROUTING -i Ethernet15 -p tcp -j DNAT -d 65.55.42.1 --dport 1024 --to-destination 20.0.0.1:6000
+iptables -t nat -A POSTROUTING -o Ethernet15 -p tcp -j SNAT -s 20.0.0.1 --sport 6000 --to-source 65.55.42.1:1024
+```
+They essentially tells the kernel to do the DNAT port translation for any incoming packets, and the SNAT port translation for the outgoing packets.
+
+If there are any ACL to NAT pool bindings configured, the NatMgrd listens to the notifications from the ACL tables and the ACL Rule configuration tables. On start-up, the NatMgrd queries the ACL Configuration for the ACLs bound to NAT pools. Once the ACL rules are retrieved, they are updated as iptables filtering options for the SNAT configuration.
+
+Before acting upon the Dynamic NAPT Pool configuration on an interface, NatMgrd checks with the STATE_DB that the matching outside IP interface is created in the system (state == ok).
+
+For the NAT_POOL configuration as below that is configured to match an ACL,
+```
+NAT_POOL|pool1
+    "nat_ip": 65.55.42.1 
+    "nat_port": 1024-65535 
+
+INTERFACE|Ethernet15|65.55.42.1/24
+...
+
+ACL_TABLE|10
+    "stage": "INGRESS", 
+    "type": "L3", 
+    "policy_desc": "nat-acl", 
+    "ports": "Vlan2000"
+
+ACL_RULE|10|1
+    "priority": "20", 
+    "src_ip": "20.0.1.0/24",
+    "packet_action": "do_not_nat"
+
+ACL_RULE|10|2
+    "priority": "10", 
+    "src_ip": "20.0.0.0/16", 
+    "packet_action": "forward"
+
+NAT_BINDINGS|nat1    
+    "access_list": "10"
+    "nat_pool": "pool1
+    "nat_type": "snat"
+
+PORT|Ethernet15
+    "nat_zone": 1
+  
```

Comment:
For dynamic source NAT, the public IP to translate to may not be a physical interface address.  It can be a loopback interface IP.  Please support this.  In the sai pipeline as well as in the spec here, it's stated that outgoing zone-id is from the next-hop interface from the L3 route lookup of the destination.  How to derive the zone-id when the public IP is a loopback IP interface address?  Thanks.

Reply:
Using loopback IP as public IP makes sense. It will be supported. Will update the spec to mention about it.
The outgoing zone-id is not affected by the Source Public IP used. The next-hop interface for the L3 route lookup to an external destination would be physical interface facing outbound.